docs/developer/source-versioning-2024-08-30.md

```diff
@@ -0,0 +1,183 @@
+    "database_name":      DatabaseNameSchema("table", false),
+    "source": IdentifierSchema(IdentifierSchemaParams{
+        Elem:        "source",
+        Description: "The source this table is created from.",
+        Required:    true,
+        ForceNew:    true,
+    }),
+    "upstream_name": {
+        Type:        schema.TypeString,
+        Required:    true,
+        ForceNew:    true,
+        Description: "The name of the table in the upstream database.",
+    },
+    "upstream_schema_name": {
+        Type:        schema.TypeString,
+        Optional:    true,
+        ForceNew:    true,
+        Description: "The schema of the table in the upstream database.",
+    },
+    "text_columns": {
+        Description: "Columns to be decoded as text.",
+        Type:        schema.TypeList,
+        Elem:        &schema.Schema{Type: schema.TypeString},
+        Optional:    true,
+        ForceNew:    true,
+    },
+    // ... other fields as needed ...
+}
+```
+
+### 4.2 Resource Implementation
+
+#### 4.2.1 Table Source Table Resource
+
+Implement CRUD operations for the new `materialize_source_table` resource:
+
+```go
+func SourceTable() *schema.Resource {
+    return &schema.Resource{
+        CreateContext: sourceTableCreate,
+        ReadContext:   sourceTableRead,
+        UpdateContext: sourceTableUpdate,
+        DeleteContext: sourceTableDelete,
+        Importer: &schema.ResourceImporter{
+            StateContext: schema.ImportStatePassthroughContext,
+        },
+        Schema: sourceTableSchema,
+    }
+}
+```
+
+The `CreateContext` function will use the new SQL syntax:
+
+```sql
+CREATE TABLE <database_name>.<schema_name>.<name> FROM SOURCE <source_name> (REFERENCE = <upstream name>) WITH (TEXT COLUMNS = (..), ..)
+```
+
+#### 4.2.2 Update Existing Source Resources
+
+Modify the CRUD operations for existing source resources to handle the deprecation of subsource-related fields:
+
+- In `Create` and `Update` operations, if the deprecated `table` field is used, log a warning message advising users to migrate to the new `materialize_source_table` resource.
+- In `Read` operations, continue to populate the `table` field if it exists in the state, but also log a deprecation warning.
+
+### 4.3 Migration Strategy
+
+We will not create separate resources with a v2 suffix for sources. Instead, we'll use a gradual migration approach:
+
+1. Deprecate the `table` field in existing source resources.
+2. Introduce the new `materialize_source_table` resource.
+3. Allow both old and new configurations to coexist during a transition period.
+
+This approach allows users to migrate their configurations gradually:
+
+- Existing sources can still be created and managed.
+- New tables (formerly subsources) will be created as separate `materialize_source_table` resources.
+- Users can migrate their configurations at their own pace by replacing `table` blocks with `materialize_source_table` resources.
```

Comment:
this makes sense to me to for allowing users to opt-in to the migration on their own accord!

We've also discussed implementing a catalog migration on the MZ side to automatically migrate all existing sources/subsources over to the new model, and in that case we would alter the `CREATE SOURCE` statement to remove the deprecated `table` field options. If the user reruns terraform after that migration it would attempt to re-introduce the `table` field unless the user was proactive in updating their terraform config in coordination with that migration. Do we think there'd be a way to detect this situation? Perhaps if the `table` field is already empty and the `enable_create_table_from_source ` feature-flag is enabled we just ignore the terraform config for that option and emit a warning?

Reply:
Yeah, in general these migrations are so hairy when they involve stuff happening inside Materialize and stuff happening with Terraform. 

I don't know that there will be much value in teaching the Terraform configuration to ignore the `table` field once the migration occurs. There's no guarantee that users will upgrade to a version of the Terraform provider that has the special handling for `table` before we do the catalog migration. And the catalog migration will happen behind Terraform's back, which is pretty sketchy—someone who thought they had everything properly tracked in Terraform will suddenly find that is not the case.

I think our best bet here may be to support the two syntaxes indefinitely, but disable the old syntax for _new_ sources as soon as we can pull off. That will prevent the problem from getting any worse. Then, after a couple months, hopefully there are few enough sources remaining that use the old syntax that we can go customer by customer and source by source and figure out if they're using Terraform to manage the source or not. If they're using Terraform, I think we're better off having field eng work with them to get through the upgrade manually than trying to do anything automatic.

Reply:
> I think our best bet here may be to support the two syntaxes indefinitely, but disable the old syntax for new sources as soon as we can pull off.

This sounds very good to me!

>  I think we're better off having field eng work with them to get through the upgrade manually than trying to do anything automatic.

I will try to prepare a step by step migrations guide on how to switch from the old to the new syntax on the Terraform side.